### PR TITLE
Let `ImbalancedSampler` accept `torch.Tensor` as input

### PR DESCRIPTION
**Pull request by EdisonLeeeee**
Hi, I modified `torch_geometric.loader.ImbalancedSampler` to accept `torch.Tensor` object, i.e., the class distribution as input. Now it is supported to use the class labels to initialize the sampler, as like:
```python
from torch_geometric.loader import NeighborLoader, ImbalancedSampler

sampler = ImbalancedSampler(data.y, input_nodes=data.train_mask)
loader = NeighborLoader(data.y, input_nodes=data.train_mask,
                        batch_size=64, num_neighbors=[-1, -1],
                        sampler=sampler, ...)    
                                sampler=sampler, ...)  
```
And the issue in #5108 should be fixed with minimal adaptation:
```python
from torch_geometric.datasets import DBLP
from torch_geometric.loader import ImbalancedSampler, NeighborLoader

dataset = DBLP(root='~/data/pygdata/dblp')
data = dataset[0]

# NOTE: input the node labels instead of the dataset `data['author']`
sampler = ImbalancedSampler(data['author'].y, input_nodes=data['author'].train_mask) 

train_loader = NeighborLoader(data, input_nodes=('author', data['author'].train_mask),
                              num_neighbors=[10, 10], 
                              sampler=sampler, batch_size=20)
```

### CODE REVIEW

**Review by EdisonLeeeee — COMMENTED** [review]
Fix several typos

File: torch_geometric/loader/imbalanced_sampler.py [hunk 1]
```diff
@@ -60,6 +72,10 @@ def __init__(
             assert dataset.num_nodes == y.numel()
             y = y[input_nodes] if input_nodes is not None else y
 
+        elif isinstance(dataset, Tensor):
+            y = dataset.view(-1)
+            y = y[input_nodes] if input_nodes is not None else y
```
Comment (Padarn) [inline torch_geometric/loader/imbalanced_sampler.py:77]: I find this condition a bit confusing - why would someone want to provide both?

File: test/loader/test_imbalanced_sampler.py [hunk 2]
```diff
@@ -41,6 +41,7 @@ def test_neighbor_loader_with_imbalanced_sampler():
     edge_index = torch.empty((2, 0), dtype=torch.long)
     data = Data(edge_index=edge_index, y=y, num_nodes=y.size(0))
 
+    # test with `torch_geometric.data.Data` instance as input
```
Comment (Padarn) [inline test/loader/test_imbalanced_sampler.py:44]: Do we need this comment?

File: torch_geometric/loader/imbalanced_sampler.py [hunk 3]
```diff
@@ -33,10 +33,22 @@ class distribution.
                                 batch_size=64, num_neighbors=[-1, -1],
                                 sampler=sampler, ...)
 
+    It is also supported to accept a :class:`torch.Tensor` object,
+    which is the node label that represents the class distribution:
```
Comment (Padarn) [inline torch_geometric/loader/imbalanced_sampler.py:37]:
```suggestion
    which has the node label distribution that we wish to sample with:
```
Is this accurate? (No need to accept this change, but the current description was confusing for me)

I would have thought you'd directly want to provide weights in this case.

File: test/loader/test_imbalanced_sampler.py (hunk 2 above)
Comment (EdisonLeeeee) [reply test/loader/test_imbalanced_sampler.py:44]: Ok, I'll remove this comment.

File: torch_geometric/loader/imbalanced_sampler.py (hunk 1 above)
Comment (EdisonLeeeee) [reply torch_geometric/loader/imbalanced_sampler.py:77]: I'm also a bit confused about this. Indeed, is it really necessary to support so many types of input for this variable `dataset`? I would suggest to support only `torch.Tensor` (labels) or `List[torch.Tensor]` ( a list of labels) as input.

```python
class ImbalancedSampler(torch.utils.data.WeightedRandomSampler):
    def __init__(
        self,
        y: Union[Tensor, List[Tensor]],
        num_samples: Optional[int] = None,
    ):

        if isinstance(y, Tensor):
            y = y.view(-1)
        else:
            if isinstance(y[0], Tensor):
                y = torch.cat(y, dim=0).view(-1)
            else:
                y = torch.tensor(y).view(-1)

    assert y.dtype == torch.long  # Require classification.

    num_samples = y.numel() if num_samples is None else num_samples

    class_weight = 1. / y.bincount()
    weight = class_weight[y]

    return super().__init__(weight, num_samples, replacement=True)
```
Users need to specify themselves which labels should be inputted
```python
from torch_geometric.loader import ImbalancedSampler

sampler = ImbalancedSampler(data.y[data.train_mask]) 
```

This may result in some changes for relevant users. I'm not so sure. Could you give me some suggestions?

Comment (Padarn) [reply torch_geometric/loader/imbalanced_sampler.py:77]: Yeah, seems fairly heavily used so maybe best not to break the current contract. You could define a new class for  this approach... probably need @rusty1s for a better opinion on this

Comment (rusty1s) [reply torch_geometric/loader/imbalanced_sampler.py:77]: I personally would like the change of only inputting a `Tensor` which results in s cleaner interface but has the trade-off that users need to first collect their label tensor. I think we can leave the interface as it is for now, and add a deprecation warning for other inputs than tensors. We would also need to update examples. Happy to also do it in a follow-up PR if this is too complicated right now.

**Review by rusty1s — APPROVED** [review]
Thanks!
